aea/helpers/async_friendly_queue.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the implementation of AsyncFriendlyQueue."""
+import asyncio
+import queue
+import threading
+from typing import Any
+
+
+class AsyncFriendlyQueue(queue.Queue):
+    """queue.Queue with async_get method."""
+
+    def __init__(self, *args, **kwargs):
+        """Init queue."""
+        super().__init__(*args, **kwargs)
+        self._lock = threading.Lock()  # to cover _waiters access
+        self._non_empty_waiters = []
+
+    def put(self, item: Any, *args, **kwargs) -> None:
+        """
+        Put an item into the queue.
+
+        :param item: item to put in the queue
+        :param args, kwargs: similar to queue.Queue.put
+        """
+        super().put(item, *args, **kwargs)
+        if self._non_empty_waiters:
+            waiter = self._non_empty_waiters.pop(0)
+            waiter._loop.call_soon_threadsafe(waiter.set_result, True)
+
+    def get(self, *args, **kwargs) -> Any:
+        """
+        Get an item into the queue.
+
+        :param args, kwargs: similar to queue.Queue.get
+        """
+        return super().get(*args, **kwargs)
+
+    async def async_wait(self) -> None:
+        """
+        Wait an item appears in the queue
+
+        :return: None
+        """
+        if not self.empty():
+            return
+        waiter = asyncio.Future()  # type: ignore
+        self._non_empty_waiters.append(waiter)
+        try:
+            await waiter
+        finally:
+            try:
+                self._non_empty_waiters.remove(waiter)
+            except ValueError:
+                pass
+
+    async def async_get(self) -> Any:
```

Comment:
maybe not now, but eventually: support for `block` and `timeout` parameters?

Reply:
`block` and `timeout` is redundant for coroutines:
* non block should be done by sync implementation and should be called instead of async block=False,
* timeout is done by `wait_for` with timeout option